planner/core/rule_predicate_push_down.go

```diff
@@ -726,3 +742,123 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
+	if len(IdxCols) == 0 {
+		return conds, nil
+	}
+
+	adder := &exprPrefixAdder{
+		sctx:      sc,
+		OrigConds: conds,
+		cols:      IdxCols,
+		lengths:   IdxColLens,
+	}
+
+	return adder.addExprPrefix4ShardIndex()
+}
+
+// AddExprPrefix4ShardIndex
+// if original condition is a LogicOr expression, such as `WHERE a = 1 OR a = 10`,
+// call the function AddExprPrefix4DNFCond to add prefix expression tidb_shard(a) = xxx for shard index.
+// Otherwise, if the condition is  `WHERE a = 1`, `WHERE a = 1 AND b = 10`, `WHERE a IN (1, 2, 3)`......,
+// call the function AddExprPrefix4CNFCond to add prefix expression for shard index.
+func (adder *exprPrefixAdder) addExprPrefix4ShardIndex() ([]expression.Expression, error) {
+	if len(adder.OrigConds) == 1 {
+		if sf, ok := adder.OrigConds[0].(*expression.ScalarFunction); ok && sf.FuncName.L == ast.LogicOr {
+			return adder.addExprPrefix4DNFCond(sf)
+		}
+	}
+	return adder.addExprPrefix4CNFCond(adder.OrigConds)
+}
+
+// AddExprPrefix4CNFCond
+// add the prefix expression for CNF condition, e.g. `WHERE a = 1`, `WHERE a = 1 AND b = 10`, ......
+// @param[in] conds        the original condtion of the datasoure. e.g. `WHERE t1.a = 1 AND t1.b = 10 AND t2.a = 20`.
+//                         if current datasource is `t1`, conds is {t1.a = 1, t1.b = 10}. if current datasource is
+//                         `t2`, conds is {t2.a = 20}
+// @return  -     the new condition after adding expression prefix
+func (adder *exprPrefixAdder) addExprPrefix4CNFCond(conds []expression.Expression) ([]expression.Expression, error) {
+
+	newCondtionds, err := ranger.AddExpr4EqAndInCondition(adder.sctx,
+		conds, adder.cols)
+
+	return newCondtionds, err
+}
+
+// AddExprPrefix4DNFCond
+// add the prefix expression for DNF condition, e.g. `WHERE a = 1 OR a = 10`, ......
+// The condition returned is `WHERE (tidb_shard(a) = 214 AND a = 1) OR (tidb_shard(a) = 142 AND a = 10)`
+// @param[in] condition    the original condtion of the datasoure. e.g. `WHERE a = 1 OR a = 10`.
+//                          condtion is `a = 1 OR a = 10`
+// @return 	 -          the new condition after adding expression prefix. It's still a LogicOr expression.
+func (adder *exprPrefixAdder) addExprPrefix4DNFCond(condition *expression.ScalarFunction) ([]expression.Expression, error) {
+
+	var err error
+	dnfItems := expression.FlattenDNFConditions(condition)
+	newAccessItems := make([]expression.Expression, 0, len(dnfItems))
+
+	for _, item := range dnfItems {
+		if sf, ok := item.(*expression.ScalarFunction); ok {
+			var accesses []expression.Expression
+			if sf.FuncName.L == ast.LogicAnd {
+				cnfItems := expression.FlattenCNFConditions(sf)
+				accesses, err = adder.addExprPrefix4CNFCond(cnfItems)
+				if err != nil {
+					return []expression.Expression{condition}, err
+				}
+				newAccessItems = append(newAccessItems, expression.ComposeCNFCondition(adder.sctx, accesses...))
+			} else if sf.FuncName.L == ast.EQ || sf.FuncName.L == ast.In {
```

Comment:
The leaf of the DNF can still be a CNF(And clause), so we may need to call `addExprPrefix4CNFCond`.

Reply:
Yes, I declared it in the doc https://github.com/pingcap/tidb/pull/31260/files#diff-5278acfa3ac95c90f2f08cb5ab1911b3892253a52e6d2fc5b172d42f47befce6R393
We scatter the hot index as simple as possible at first, It was asked to cover the certain scenes.